### PR TITLE
client/http, retry: support backoffer in HTTP client

### PR DESCRIPTION
**Pull request by JmPotato**
### What problem does this PR solve?

Issue Number: ref #7300, #7673.

### What is changed and how does it work?

```commit-message
Support backoffer in HTTP client.
```

### Check List

Tests

- Unit test
- Integration test

### Release note

```release-note
None.
```

### CODE REVIEW

File: client/http/client.go [hunk 1]
```diff
@@ -413,6 +421,13 @@ func (c *client) WithRespHandler(
 	return &newClient
 }
 
+// WithBackoffer sets and returns a new client with the given backoffer.
+func (c *client) WithBackoffer(bo *retry.Backoffer) Client {
```
Comment (HuSharp) [inline client/http/client.go:425]: Do we need to use a default backoff?

Comment (JmPotato) [reply client/http/client.go:425]: We could leave this as an option to the caller. If the caller needs one, it could initialize a backoffer by itself.

Comment (HuSharp) [reply client/http/client.go:425]: get it. Please add a test in client_test

File: client/http/interface.go [hunk 2]
```diff
@@ -92,6 +93,8 @@ type Client interface {
 	// Additionally, it is important for the caller to handle the content of the response body properly
 	// in order to ensure that it can be read and marshaled correctly into `res`.
 	WithRespHandler(func(resp *http.Response, res interface{}) error) Client
+	// WithBackoffer sets and returns a new client with the given backoffer.
+	WithBackoffer(*retry.Backoffer) Client
```
Comment (lance6716) [inline client/http/interface.go:97]: I prefer caller use `ClientOption` in `NewClient` to set it.

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: client/http/interface.go (hunk 2 above)
Comment (JmPotato) [reply client/http/interface.go:97]: Using `WithBackoffer` here allows us to flexibly use different `Backoffer`s for different scenarios and interfaces like:

```go
cli := NewClient()
cli.WithBackoffer(bo1).GetRegion()
cli.WithBackoffer(bo2).GetStore()
```

If we use`ClientOption` for initialization, we may need to create a new `Client` every time we want to use a different `Backoffer`.

**Review by lance6716 — APPROVED** [review]

File: client/retry/backoff.go [hunk 3]
```diff
@@ -48,6 +49,8 @@ func (bo *Backoffer) Exec(
 		after := time.NewTimer(currentInterval)
 		select {
 		case <-ctx.Done():
+			after.Stop()
+			return errors.Trace(ctx.Err())
```
Comment (HuSharp) [inline client/retry/backoff.go:53]: Do we need to reset the backoffer here?

Comment (JmPotato) [reply client/retry/backoff.go:54]: Nice catch. Fixed.

**Review by HuSharp — APPROVED** [review]

File: client/http/client.go [hunk 4]
```diff
@@ -287,7 +293,8 @@ func (ci *clientInner) updateMembersInfo(ctx context.Context) {
 		WithName(getMembersName).
 		WithURI(membersPrefix).
 		WithMethod(http.MethodGet).
-		WithResp(&membersInfo))
+		WithResp(&membersInfo).
+		WithBackoffer(nil))
```
Comment (rleungx) [inline client/http/client.go:297]: If no need to back off, how about just removing WithBackoffer?

File: client/retry/backoff.go [hunk 5]
```diff
@@ -18,65 +18,103 @@ import (
...
-		defer after.Stop()
+) (err error) {
+	defer bo.resetBackoff()
+	for {
+		err = fn()
+		if err == nil {
+			break
+		}
+		currentInterval := bo.nextInterval()
+		after := time.NewTimer(currentInterval)
```
Comment (rleungx) [inline client/retry/backoff.go:50]: Is it necessary to create a timer every time?

**Review by rleungx — APPROVED** [review]

**Pull request merged**